### DISCUSSION TITLE
Modelling a custom PFR model using DAEs

### DISCUSSION THREAD

**Discussion by abhi24x7** [IDAES/idaes-pse#1355] [Q&A]
Dear IDAES community,

I just started using the framework a week ago and I would like to thank all the concerned developers for working on the tutorials and examples, they really gave me insights about how the platform works. It was a really good guide.

Now, through this thread, I am seeking your help to make a **custom packed bed PFR model** with my own Mass & Energy balance equations and rate equations (which are not simple Arrhenius model), I just want to know how can I declare the same?

Thank you very much!

**Comment by andrewlee94** [comment]
Hi @abhi24x7, welcome to IDAES. It seems there are two parts to your question, and it sounds like you are trying to do a fairly custom model.

Firstly, on the topic of the reactor itself. It is possible to use the core PFR model and turn off the mass, energy and/or pressure balances if you desire, but at that point you might be better off asking if it is better to use a completely custom model. Depending on how you want to write the reaction model, you might also need to change how the extent of reaction is defined - the PFR uses `extent = area*rate` (i.e. a volumetric definition of reaction rate). To turn off the material balances, you can do `m.fs.pfr = PFR(material_balance_type=MaterialBalanceType.useDefault)` (and similar for the energy and momentum balances). However, before you start working on this, it might be a good idea to discuss what it is you want to do in more detail so we can give some more detailed suggestions on how to go about it (specifically, what is it you want to customize about the mass and energy balances).

On the topic of the reactions, you can write your own custom reaction package. A simple example can be found here: https://github.com/IDAES/idaes-pse/blob/main/idaes/models/properties/examples/saponification_reactions.py. This uses a simple Arrhenius power law model, but you can replace those equations with those of your choice. As long as the final rate is on a volumetric basis (e.g. mol/m^3), it will fit into the core models.

**Reply by abhi24x7** [reply]
Dear @andrewlee94!
Thank you very much for replying to my problem. Sure I shall provide you a more clear version of what I am planning to implement using IDAES framework.

**About the Reactor**
I want to model a **packed bed tubular reactor** to model a certain reaction which has a **non-reaction Arrhenius type** rate equation which includes also involves the partial pressures of the components and the property of catalyst (like sphericity and density). As yet, the reaction property package is defined based by taking help from the discussion forum. 

Just to start  and keeping it simple, 
I model the reactor as a simple PFR that is already there, in the current flowsheet I just have the FEED block, HEATER block, the PFR and finally the PRODUCT block. Uptil here the flowsheet converges fine and the results seem okay! But now upon adding two more unit operations like the adding the COMPRESSOR and two more HEATERS, the flowsheet fails to converge. And upon using the .display( ) method to check about what seems to be the issue, I see that my rate equation throws a residual error in the equality (==0) constraint section. This doesn't happen in the previous more simpler model as I rechecked my definitions of rate equation, they seem to work fine. However it isn't the case in the second flowsheet as still it doesn't converge, I tried changing the P and T conditions of the compressor and heater system, just to understand what happens, but still it doesn't alter much and the flowsheet still doesn't converge.

Suggestions/tips regarding the above issue would be highly appreciated.

Regarding the custom DAE, I have previously solved coupled ODE systems (Mass balance, Energy and Pressure) of packed reactor model using Python _solve_ivp_ methods. I was wondering if I can define those ODE equations in the similar way in IDAES. I tried looking up the repository regarding this if I can find an example, but I couldn't find any but it can also be the case that I missed out something. Could you please aid me in this aspect as well?

Thank you very much in advance!

**Reply by andrewlee94** [reply]
OK, so the first thing to do here is work out why your model does not converge. When you say you added two new unit models to the process and it fails to converge, how did you go about initializing and solving the model? Does your flowsheet contain any recycle loops (I suspect not as you did not mention splitters or mixers)?
 
1. If you have no recycle loops then each unit in the flowsheet should be able to be initialized sequentially, and if the flowsheet then fails to solve it generally indicates poor scaling. I suspect this might be the issue you are running into here, especially as the simpler model can solve but the more complex one cannot.
2. If you do have recycle streams, you will need to be careful as it is very easy to make the overall system of equations degenerate.

Also, you can try applying the IDAES Diagnostics Toolbox to see what it tells you about the model. You can apply this tool to either individual unit models (start with your reactor) or the full flowsheet, and it can help tell you if you have singularities or degeneracies in the models.

https://idaes-examples.readthedocs.io/en/latest/docs/diagnostics/diagnostics_toolbox_doc.html
https://idaes-pse.readthedocs.io/en/stable/explanations/model_diagnostics/index.html

Regarding DAEs, we use the `pyomo.dae` tools for everything we do. We write the equations in their DAE form, and then use the DAE transformation to discretize these for us and then just send the model to a normal NLP solver (i.e. IPOPT most of the time). I do not know about the `solve_ivp` method or what it does.

**Reply by abhi24x7** [reply]
Dear @andrewlee94,

Thank you very much, for the above reply. I actually found the issue, it is in the customized rate equation I had defined. I tried to work around it, but not proceeding much forward.

So I refer back to the IDAES examples, more specifically the **methanol flowsheet** example. I see that in the example under the Python script _methanol_reactions.py_ a basic example of an Arrhenius based rate model is given, how do I proceed in case of more complex rate equation like as the one given below:
![image](https://github.com/IDAES/idaes-pse/assets/66618073/e0e11892-8c72-490f-be94-ae1f9f529ef2)

I searched the repository, but I couldn't find a file I could refer to. Do you have any suggestions on how to proceed?

Thank you very much in advance!

**Reply by andrewlee94** [reply]
The methanol synthesis example is not what you need for this - that uses the modular reaction framework which only has libraries for simpler reactions (you can add more complex reactions of your own, but it is probably not worth the effort learning how to do it).

This is a better example of what you need: https://github.com/IDAES/idaes-pse/blob/main/idaes/models/properties/examples/saponification_reactions.py
This is an example of how to write your own reaction package directly (the modular framework is basically setting this up for you using a library of pre-built reaction forms).

**Reply by abhi24x7** [reply]
Dear @andrewlee94,

Thank you very much for your response. I was finally able to follow and setup the custom reaction package based from your support.

**Reply by abhi24x7** [reply]
Hello @andrewlee94,

Thank you for your previous response.

I was able to add the custom reaction rates into the PFR block (named R101) by referring to the example you provided above. When I run the model, a very strange thing occurs, when I first initialize the PFR by running the cell containing the following lines, 

`m.fs.R101.initialize(outlvl=idaeslogger.DEBUG, optarg = {"tol": 1e-06, "max_iter": 2000})`
`propagate_state(arc=m.fs.s06)`

It provides a RunTime log as provided in the [PFR_R101_Initialization_ErrorLog.txt](https://github.com/IDAES/idaes-pse/files/14816986/PFR_R101_Initialization_ErrorLog.txt.txt) text file attached . And an _InitializationError_ as shown in the snapshot below:
![image](https://github.com/IDAES/idaes-pse/assets/66618073/b93d1844-fcae-4678-918d-3b42db6bfd91)


**However**, when I re-run the cell, the PFR initializes successfully!

I fail to understand why is this happening and what can I do to fix it. Because I think the situation which I described above may give rise to problems in the future.

Thank you and looking forward to your reply!

**Reply by andrewlee94** [reply]
@abhi24x7 That is actually not that unusual, although it indicates that your model is poorly scaled. In short, IPOPT does some internal scaling of your model based off the initial state of the model and then tries to solve. In the first pass, the initial state is rather poor so the scaling is also poor. This means that whilst it get close to the actual solution, the scaling means that it is unable to resolve all constraints to tolerance. When you re-run the solver though, it now uses the near-correct answer to do the scaling and gets much better results.

To address the underlying numerical issues, try using the DiagnsoticsToolbox: https://idaes-pse.readthedocs.io/en/stable/explanations/model_diagnostics/index.html. It will help identify the poorly scaled variables and constraints, along with any other issues your model might have.

**Reply by abhi24x7** [reply]
Dear @andrewlee94,

Thank you for your suggestion regarding model scaling. 
And your suggestion does make sense, even in the `methanol_synthesis.py` example which I have been using as a reference, there is a method call for scaling called `scale_flowsheet(m)`.  I shall definitely try out the `DiagnosticToolbox` and attempt the scaling, regarding which I have a few queries.

1. What does scaling do exactly, is it somewhat like normalization? As from what I understand after reading the documentation and your previous reply it is a technique used to ease numerical stiffness to improve convergence, is my understanding correct?
2. Do I scale only at the unit level like how its is shown in `scale_flowsheet(m)` , or also at the property package as in the `build_model()` method in `methanol_synthesis.py` example as shown below
![image](https://github.com/IDAES/idaes-pse/assets/66618073/20247672-7b09-4054-93ef-7404dfd071af)
4. Could you please provide me some advice on how to choose the scaling factors, can I use the ones in the `methanol_synthesis.py` example or are they problem specific?

Thank you very much and looking forward to learn something from your reply!

**Reply by andrewlee94** [reply]
Scaling is a fairly complex topic, and not something I am the best person to teach either. The goal is to improve convergence, although it is more than just stiffness. Scaling is also made up of three parts; variable scaling, constraint residual scaling and Jacobian (partial derivative scaling) in roughly increasing order of importance.

Probably the simplest way to explain it is in terms of the condition number of the problem (shown in the numerical diagnostics); you can view this as the gain (amplification) of errors in the model, and the bigger the number the more it will amplify errors. The first goal of scaling is to bring the condition number down (at a minimum, <1e10).

For how to implement scaling, start here: https://pyomo.readthedocs.io/en/stable/model_transformations/scaling.html. I would generally avoid looking at the IDAES implementation of scaling, as it has a number of limitations that are hard to explain (and it is rather opaque to the user). We are in fact working on completely overhauling the IDAES scaling tools this year.

Finally, for how to pick good scaling factors; there is no simple (or even "good") answer. Start by looking at the rows and columns (constraints and variables) in your model with extreme Jacobian values and attempt to bring them to a value between 1 and 10 (the diagnostics will identify these for you). Also, an important final note: good scaling depends on the solution, not the current state so you should ideal set scaling factors based on the answer. However, you obviously would not be solving the model if you already knew the answer, so try to take your best guess.

**Reply by abhi24x7** [reply]
Dear @andrewlee94,

Thank you for your above reply.

In the DiagnosticToolbox, when I run the method `report_numerical_issues()` it gives me an Runtime error stating, _Pynumero not available_.   Is Pynumero a IDAES's package or PyOMO's. I couldn't install it even using `pip install`  is there some other way to install this?

Thank you very much and looking forward to your reply!

**Reply by andrewlee94** [reply]
PyNumero is a part of Pyomo, and needs to be installed separately. See https://pyomo.readthedocs.io/en/stable/contributed_packages/pynumero/installation.html

**Reply by abhi24x7** [reply]
Dear @andrewlee94,

Thank you for your previous reply. 
I have 2 queries to clarify,

1. Installation of Pynumero
I completely understand that this discussion thread is solely meant for issues related to IDAES, but installation of Pynumero is crucial as one of the methods in IDAES needs it, hence daring to ask here.
Based on your previous reply I went through the [link](https://pyomo.readthedocs.io/en/stable/contributed_packages/pynumero/installation.html) to install Pynumero, but I fail to understand how the methods described in the link applicable are to me. As I run IDAES, Pyomo on Anaconda environment (like I installed both idaes-pse and pyomo using `pip install` in Anaconda Prompt). More specifically I also tried to install Pynumero using the methods described [here ](https://anaconda.org/conda-forge/pynumero_libraries), I get the following error:

![image](https://github.com/IDAES/idaes-pse/assets/66618073/900f2c0f-1c87-4514-9bdf-13d9a283b347)

Do you have any suggestions about how can I go about this, to make things worse the Pynumero is not a mainstream library, hence there are just a couple of methods to install (which I attempted but was unsuccessful).


2. A small doubt about the scaling factors in  HDA Flowsheet example.
I was trying to understand how scaling is done from a Pyomo perspective in IDAES and I learnt that the method in which scaling has been done is different from one flowsheet example to another. And more specifically I have a small question about HDA flowsheet example, as documented in https://idaes.github.io/examples-pse/latest/Examples/Flowsheets/HDA_flowsheet_with_distillation_solution_doc.html, here in one section of the jupyter notebook.

![image](https://github.com/IDAES/idaes-pse/assets/66618073/babfc4a5-d1fd-4c91-9088-940127a2f020)

More specifically the second part of the code about _scaling factors for the remaining variables and all constraints_, does it mean that running these lines of code scales all the variables and constraints (including all the user defined as well?) involved with the particular unit block? Moreover in this section, the method `iscale.calculate_scaling_factors()` was used and in the previous `iscale.set_scaling_factor()` was used.
From what I understand `iscale.calculate_scaling_factors()` automatically calculates the scaling factors for all the variables and constraints involved with the unit and sets it, however `iscale.set_scaling_factor()`  sets the scaling factor which the user passes. Is my understanding correct? 

Thank you and looking forward to your reply!

**Reply by abhi24x7** [reply]
@andrewlee94,

I was going through the IDAES documentation to understand more about scaling and I came to know about the **[ScalingToolbox](https://idaes-pse.readthedocs.io/en/stable/reference_guides/scaling.html#id1)**. While reading the documentation I came across this method called `idaes.core.util.scaling.report_scaling_issues()`, when I try using this method on one of the heater blocks (H101) by using the method as `iscale.report_scaling_issues(m.fs.H101)` I get the following error [Scaling_H101.txt](https://github.com/IDAES/idaes-pse/files/14839190/Scaling_H101.txt).

![image](https://github.com/IDAES/idaes-pse/assets/66618073/a48097a8-8c13-4f68-b593-30a4027029ee)

So from what I understand about these results is that, all the variables and constraints indicated above, I have to bring their values in between 1 and 10 by implementing scaling factors. Is  my understanding correct?

However, another potential issue I can see now is that, I am doing the scaling before the initialization of the unit blocks. But at this stage certain variables in other components such as the PFR that uses a custom defined rate equation, in there the rate some reaction rate terms go beyond 1e7 (i know this as I checked the values of these rates after solving the poorly scaled model). So to set the scaling factors of such terms I first get an idea of the order of magnitude by solving the poorly scaled model and then start setting the scaling factors accordingly before the initialization such that the magnitude of these terms lie between 1 and 10. and to verify the results of the scaling by solving the model. Is this approach correct?

Please correct me if I have misunderstood something.

Thank you and looking forward to your reply.

**Reply by blnicho** [reply]
@abhi24x7 PyNumero is distributed with the IDAES extensions. So assuming you ran `idaes get-extensions` during your installation of IDAES you should already have it. See the full list of what's included here: https://idaes-pse.readthedocs.io/en/latest/tutorials/getting_started/binaries.html

**Reply by abhi24x7** [reply]
Dear @blnicho ,

Thank you for the steps you suggested and Pynumero now works! or atleast the `DiagnosticToolbox` function  `report_numerical_issues()` now works and gives results.

**Reply by abhi24x7** [reply]
Hello,

I need some advice about the procedure I am following to scale the model.

The procedure I follow is as follows:
I have an unscaled model, I first initialize it and then solve it. Post solving it, I use the `DiagnosticsToolbox` and I use the `report_numerical_issues()` method to check for the numerical issues, here it prints a Jacobian number of 7.575E+20 (which confirms the model is not scaled well)
Post which I use the `display_constraints_with_extreme_jacobians()` ([Constraint extreme jacobian output.txt](https://github.com/IDAES/idaes-pse/files/14849883/Constraint.extreme.jacobian.txt)) and `display_variables_with_extreme_jacobians()` ([Variable extreme jacobian.txt](https://github.com/IDAES/idaes-pse/files/14849897/Variable.extreme.jacobian.txt)) methods to see which constraints and variables are reaching extreme values. 

After this I run the method `report_scaling_issues()` from the `ScalingToolbox`  from which I get a list of Unscaled variables, Badly scaled variables and Unscaled constraints from the entire model. Inorder to get an idea about the magnitude of variables I use the `.value` command and let's say if the order is ~1e4, What I now have to do is apply a scaling factor of 1e-3 so that the variables come in the range of 1 to 10? Is this correct? I was also thinking of using the method `set_variable_scaling_from_current_value()` to scale the variables.

And to scale the constraints I use the `set_constraint_scaling_harmonic_magnitude()` or do I have to scale the constraints individually as they have done in the methanol_flowsheet example?

Does scaling the variables and constraints imply that, jacobians will automatically shall be scaled or do I have to scale them by using methods like `constraint_autoscale_large_jac(m)`?

Moreover, once I follow the above and initialize the model again, the solver gives out a warning as highlighted below:
![image](https://github.com/IDAES/idaes-pse/assets/66618073/01a3c763-14c0-4a26-808b-795794d0cea7)
What seems to be the issue here?

I seem to be a little stuck at the scaling hence seeking some guidance, Is the procedure I am following the right approach? Please let me know about the same, thank you very much!

**Reply by andrewlee94** [reply]
@abhi24x7 Sorry for the late reply - I was at a review meeting all last week.

1. I would generally encourage you to NOT use the IDAES scaling toolbox; it is rather opaque as to what it does (specifically the `calculate_scaling_factors` function) and it is very narrowly tailored to a specific workflow we had at the time of writing. We are working to develop a more general version at the moment, but that will not be available for many months yet. I would even ignore the method to show unscaled variables, as that really doesn't matter (what matters is the Jacobian which the diagnostics will tell you).
    * That said, some methods are useful - `set_scaling_factors` is a useful convenience tool, and the `set_constraint_scaling_harmonic_magnitude()` can be useful (but should be used with care). `constraint_autoscale_large_jac(m)` just replicates IPOPT's internal scaling routines, and you generally should not need to use this.
    * The `iscale.set_scaling_factor` function is a simple wrapper function for setting the value of a scaling factor for a given component in the scaling Suffix (`iscale.set_scaling_factor(model.var, value)` is equivalent to the Pyomo `model.scaling_factors[model.var].set_value(value)`.
    * `set_constraint_scaling_harmonic_magnitude()` can be applied to either a single constraint, or an entire model in which case it iterates over all constraints in the model. Note that this requires scaling factors for all variables in order to work, thus you should generally scale all the variables, and then scale the constraints. Note however, that this is just a heuristic method and is not guaranteed to give good scaling (but might be useful as a first guess). This gets to your question about constraint versus Jacobian scaling; Jacobian scaling is dependent on variable and constraint scaling, but the relationship is not linear. Generally, you need to set the scaling factors on the variables and constraints such that the Jacobian ends up well scaled, which can take some trial and error. However, there is no separate scaling factors for the Jacobian; there is just variable and constraint scaling.
    * `iscale.calculate_scaling_factors` does a lot of things, but first and foremost it calls a method named `model.calcualte_scaling_factors()` (very poor choice of names) to scale each model. That is - it relies on the model having implemented a method to do scaling, which is generally not true for custom models. Also, it should be noted that this method also _applies_ constraint scaling factors directly, which is very bad practice (as I said, this was tailored to a specific workflow).
    * The warning you see is because of the `calculate_scaling_factors` function/methods, which are misusing the Pyomo scaling suffix (originally Pyomo did not raise a warning about this, but got more vocal at some point). You can safely ignore this warning, but this is part of why we are working to rebuild the scaling tools.
  
3. To start, you should get familiar with the Pyomo scaling transformation and suffix, as that is how scaling is defined (IDAES just builds on these). Of important note, you should understand the use of Suffixes for setting scaling factors, the scaling transformation for creating scaled model, and the `propagate_solution` method for getting results back to the unscaled model (this is important to remember, and is easily overlooked). Also keep in mind that if you want/need to rescale the model then you need to run the scaling transformation again.
4. When using the diagnostics tools ,you will need to apply these to the scaled model to get useful answers - i.e. you need to build the model, set the scaling factors, apply the Pyomo scaling transformation to create a scaled model and then use the diagnostics to inspect the scaled model.
5. Scaling is inherently an iterative process - as you note when you first start out you do not know the magnitude of the variables at the solution thus it is often hard to get good scaling factors. You need to make your best guess, try to solve the model, and then refine once you have a better solution. This also applies each time you try to solve the model, as the variable magnitudes will keep changing.

That is a lot to sort throguh, so I'll end here. In short, scaling is very important but is not easy and often requires trial and error.

**Reply by abhi24x7** [reply]
Dear @andrewlee94, 

Thank you for the elaborative reply and please don't be sorry, I can understand. Just to update you I have implemented the scaling factors and I have seen the improvement in the model, my reactor initializes at one-go.

I wanted to understand one small thing which I have noticed about the compressor unit which I have in my flowsheet, so this compressor raises the pressure of the gas feed from 1 bar to 20 bar, and when I used the `report()` function, I get as follows:
![image](https://github.com/IDAES/idaes-pse/assets/66618073/f3268701-8924-44ee-8e3f-c705cc8779ac)

I don't understand why the Mechanical work is in such a low order of magnitude considering I am compressing $10^3$ order of magnitude of feed. 

Another thing is, as yet in the compressor definition I use the `ThermodynamicAssumption.isothermal` as the thermodynamic assumption, which doesn't change the compressed feed temperature (is it because of this that the mechanical work is so low?).

Could you please recommend me about how can I see the change in temperature of the compressed feed?

Thank you and looking forward to your reply!

**Reply by andrewlee94** [reply]
@abhi24x7 The answer I think is fairly straight forward - if you use the isothermal assumption then the temperature will not change by definition (isotherm = same temperature). I also suspect you are using the ideal gas assumption for you gas properties - one of the biggest shortcomings of the ideal gas assumption is that enthalpy is not a function of pressure, and thus compression is "free".

If so, the issue is the assumptions you are using and the result is actually entire expected (and entirely wrong). To get meaningful values from a compressor model, you need to use a non-ideal property package, and a more rigorous thermodynamic model for the compressor (e.g. isentropic). Of course, this does add more complexity to the model as well which will liekly mean more things that need to be scaled.

One very important thing to note if you do this (and one that is easily overlooked) is to never set a concentration/mole/mass fraction to 0 (or allow this to happen). Rigorous thermodynamic models involve a number of `log(x)` terms which will fail badly if a concentration goes to zero. So, be sure to set a small positive bound and value (1e-8 is often good enough) for any concentration that approaches zero.

**Reply by abhi24x7** [reply]
Dear @andrewlee94,

Thank you for your previous suggestion and apologies for my late reply. 

I wanted to clarify, a small thing for the compressor as defined as 
![image](https://github.com/IDAES/idaes-pse/assets/66618073/c85a82c5-aaf6-4d52-a86d-11a49b080b03)

Now I am using the isentropic assumption,  the results are as follows:
![image](https://github.com/IDAES/idaes-pse/assets/66618073/bded91cb-11d0-4c69-9267-328afa25acaf)

And the inlet and outlet temperature still remains the same
![image](https://github.com/IDAES/idaes-pse/assets/66618073/d40f0d08-9a63-4b02-9f23-73c0bd0a2cef)
![image](https://github.com/IDAES/idaes-pse/assets/66618073/7bfb8121-44d8-4ea3-a4d3-c6d428407aea)


From a thermodynamic point of view, the temperature should increase under isentropic compression, so could you please help me understand this?


Secondly, 
Another approach I thought is to do the following:
calculate the Compressor discharge temperature using:
![image](https://github.com/IDAES/idaes-pse/assets/66618073/0c57008c-c4fc-49c0-b395-fbef34eae49b)

Values of Gamma are easy to get for any gas

Once we have found out discharge temp (T2), We can use the following equation to get the power consumption of the compressor:
![image](https://github.com/IDAES/idaes-pse/assets/66618073/e6fcc9a6-d4e5-4620-832d-7304517d9850)


here, k is the isentropic factor , M = molar weight of gas mix and Qm = mass flowrate of gas

To implement them in IDAES, we can manually write `Expression` on the compressors

What do you think about this approach? Please let me know if there is a simpler way to do it.

Thank you in advance!

**Reply by andrewlee94** [reply]
@abhi24x7 There are a few reasons why you could be seeing those results.

1. Firstly, make sure you are checking the solver status after the the initialization and solve steps - one common reason for seeing results like this is a solver failure that you overlooked.
2. Second, what property package/thermodynamic model are you using for the gas? As I mentioned earlier, this can affect the results with different models having different behaviour (e.g., this behaviour would be expected if you are using the ideal gas law). It might help if you could show the full code you are using.

Finally, regarding your suggestions for equation; they should not be required for most thermodynamic models as the compressor model already has equations to predict the outlet temperature and work required. The issue is going to be something about the model you are using or the solver. The equations you showed are an alternative approach however, but should would require a new variant of the compressor model to implement.

**Reply by abhi24x7** [reply]
Dear @andrewlee94,

I now defined the compressor block using a non-ideal property package and it solves the issue. Thank you for your help!

**Reply by abhi24x7** [reply]
Dear @andrewlee94,

Thank you very much for all your previous help. Is there any method through which I can generate a stream summary of simulation table to a `.csv` file? Could you please refer me to an example which does a similar thing?

Thank you in advance! Looking forward to your reply.

**Reply by andrewlee94** [reply]
@abhi24x7 For a flowsheet, you should be able to call `flowsheet.stream_table(time_point)` to get the stream table as a Pandas dataframe. You can then send this to a csv file using the Pandas `to_csv` method: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_csv.html

**Reply by abhi24x7** [reply]
Dear @andrewlee94,

Could could you please guide me on how to use `flowsheet.stream_table(time_point)`? I can't figure it out. 

Thank you in advance! Looking forward to your reply.

**Reply by andrewlee94** [reply]
`stream_table` is a method which exists on all `FlowsheetBlocks` - one you have built your flowsheet you can just call this method to get the stream table. `time_point` is primarily for dynamic simulations and lets you pick a point in the time set to show the stream table for.

E.g.
```
m = ConcreteModel()
m.fs = FlowsheetBlock()

# Add properties, unit models and Arcs
# Initialize and solve

m.fs.stream_table()
```

**Reply by abhi24x7** [reply]
Dear @andrewlee94,

Thank you so much for the previous reply! This was a really really helpful command.

**Reply by abhi24x7** [reply]
Dear @andrewlee94,

I thank you very much for all the support you provided through this forum, I really really appreciate it. It was my first time using an Equation Oriented flowsheet modelling.

**Reply by ColbyCodey** [reply]
@andrewlee94 ,
I have been dealing with similar scaling issues trying to model a combined cycle system very similar to the ngcc example. I tried using a scaling process similar to that used in the ngcc example, but I still had a large number of poorly scaled variables and constraints according to the diagnostics, so I have been trying to implement some of what you recommended in your [April 8th reply.](https://github.com/IDAES/idaes-pse/discussions/1355#discussioncomment-9046531)  

I started by solving my unscaled model (it does not converge after 200 iterations but seems close with only a few small residuals), then using some of the IDAES scaling tools to automatically scale the variables, and adding a scaling factor of 1 to any variables with a value of 0.

I then tried to use * `set_constraint_scaling_harmonic_magnitude()` on the whole model but it is running into issues. I added some code to the `_set_sf_har_mag()` function to print the component name before the `NominalValueExtractionVisitor` is called, as this is where the error occurred. More specifically, the error occurs in the `_get_nominal_value_for_sum_subexpression()` function: 
![image](https://github.com/user-attachments/assets/e1996673-90be-4a1a-9d15-205d7ecefbce)
I edited this function as well to see where the str data is coming from: 

```
def _get_nominal_value_for_sum_subexpression(self, child_nominal_values): # function edited to identify issues with str present causing error with sum
        total = 0
        for i in child_nominal_values:
            print(f"\t{i}")
            if type(i) is not str:
                total += i
        return total
```
This has allowed me to identify that certain components have child nominal values that are not numbers, and even with the change to ignore these values, these components typically have some type of error occur when evaluating expressions as well (see output below):
```
Component: fs_hrsg.ECON.delta_temperature_in_equation[0.0]
	1
	1000
	0.001
	1
	0.001
	0.018015268
	1
	1000
	0.001
	1
	H2O
	985.177782734931
	7079.99998895492
	[c:\...\anaconda3\envs\idaes-tea\lib\site-packages\idaes\models\properties\general_helmholtz\components\parameters\]
ERROR: evaluating object as numeric value: t_hp_func(0, 985.177782734931,
7079.99998895492, 0)
        (object: <class
        'pyomo.core.expr.numeric_expr.ExternalFunctionExpression'>)
    [WinError -1073741784] Windows Error 0xc0000028
```
I am just hoping to get a good rough start on scaling with this. For now, I can try and work around these problematic components and manually set the scaling. I have seen this with a custom constraint I made as well as the delta temperature in equation for the HeatExchanger model so far. If you could help me to identify if I am doing something wrong that is causing this I would greatly appreciate it. 

Thanks

**Reply by andrewlee94** [reply]
@ColbyCodey That is rather odd, and it looks like there are two separate issues here. Firstly, how strings are ending up in the expression walker, and secondly why the external function is not being evaluated correctly. Do you have a simple example that replicates the issue? That said, this tool is in the process of being deprecated so I am not sure how much effort we will put into maintaining it.

Whilst this is still in the draft stage, this PR might help you: https://github.com/IDAES/idaes-pse/pull/1429 - this is what will eventually replace the tools you are looking at. This is the draft of our new scaling toolbox, which adds some additional methods that are often better the the harmonic mean (from experience, the inverse of the maximum constraint magnitude works best). It also has some draft documentation, and a tool that will compare a few different ways of doing constraint scaling for you.

**Reply by ColbyCodey** [reply]
@andrewlee94 ,
I put together a simple model with just the HeatExchanger unit model,  initialized and solved at conditions similar to what I am working with, then went through the scaling process I talked through above: 
[set_constraint_scaling_harmonic_magnitude_issues.pdf](https://github.com/user-attachments/files/16911945/set_constraint_scaling_harmonic_magnitude_issues.pdf)
This is the same error I run into with my full model.

I'll start looking into the new scaling toolbox. Love what I am seeing so far.

**Reply by andrewlee94** [reply]
@ColbyCodey I worked out what the issue was, and if was an edge case we had overlooked when working with external functions (occasionally they will have an input that is a string). I suspect the new scaling toolbox might suffer from this same issue, and if so I will update that pull request accordingly.

As a patch, you can replace lines 1476 to 1479 in `idaes.core.util.scaling.py` with:

```
input_mag = []
    for i in child_nominal_values:
        if isinstance(i[0], str):
            input_mag.append(i[0])
        else:
            input_mag.append(self._get_nominal_value_for_sum_subexpression(i))
```

**Reply by andrewlee94** [reply]
@ColbyCodey I just opened a PR with the bug fix: https://github.com/IDAES/idaes-pse/pull/1487
